File: blind75/54_palindromic_substrings/palindromic_substrings.py
def count_substrings(s: str) -> int:
    """
    Returns the number of palindromic substrings in s.
    Uses expand around center approach for O(n²) time complexity.

    Args:
        s: Input string

    Returns:
        Number of palindromic substrings
    """
    if not s:
        return 0

    count = 0

    for i in range(len(s)):
        # Count odd-length palindromes (center at i)
        count += expand_around_center(s, i, i)

        # Count even-length palindromes (center between i and i+1)
        count += expand_around_center(s, i, i + 1)

    return count


def expand_around_center(s: str, left: int, right: int) -> int:
    """
    Expands around the center and counts palindromes.

    Args:
        s: Input string
        left: Left pointer
        right: Right pointer

    Returns:
        Number of palindromes found during expansion
    """
    count = 0

    while left >= 0 and right < len(s) and s[left] == s[right]:
        count += 1
        left -= 1
        right += 1

    return count

File: blind75/54_palindromic_substrings/palindromic_substrings.go
package palindromic_substrings

// CountSubstrings returns the number of palindromic substrings in s.
// Uses expand around center approach for O(n²) time complexity.
func CountSubstrings(s string) int {
	if len(s) == 0 {
		return 0
	}

	count := 0

	for i := 0; i < len(s); i++ {
		// Count odd-length palindromes (center at i)
		count += expandAroundCenter(s, i, i)

		// Count even-length palindromes (center between i and i+1)
		count += expandAroundCenter(s, i, i+1)
	}

	return count
}

// expandAroundCenter expands around the center and counts palindromes.
func expandAroundCenter(s string, left, right int) int {
	count := 0

	for left >= 0 && right < len(s) && s[left] == s[right] {
		count++
		left--
		right++
	}

	return count
}
